lib/lexicon.py

```diff
@@ -104,6 +107,42 @@ def from_element(cls, e):
         synt = None if not synt else synt[0]
         return Lemma(orth, phon, synt, eval, special)
 
+    def _equals(self, other: Lemma, same_order: bool = True) -> bool:
+        """
+        Check for lemma equality.
+
+        :param other: Other lemma to compare :param:`self` to.
+        :param same_order: Whether the order in the different lemma elements matters or not.
+        :return: Whether :param:`self` and :param:`other` are equal or not.
+        """
+        if same_order:
+            return (
+                self.orth == other.orth
+                and self.phon == other.phon
+                and self.special == other.special
+                and self.synt == other.synt
+                and self.eval == other.eval
+            )
+        else:
+            if self.synt is not None and other.synt is not None:
+                equal_synt = set(self.synt) == set(other.synt)
+            else:
+                equal_synt = self.synt == other.synt
+
+            return (
+                set(self.orth) == set(other.orth)
+                and set(self.phon) == set(other.phon)
+                and self.special == other.special
+                and equal_synt
+                and set(itertools.chain(*self.eval)) == set(itertools.chain(*other.eval))
+            )
+
+    def __eq__(self, other: Lemma) -> bool:
+        return self._equals(other, same_order=False)
```

Comment:
Is there a reason you want `same_order=False` here?

It probably depends on the use case of `__eq__` here (what actually is your use case?) but it feels kind of arbitrary.

Reply:
This should be discussed, but in my view a lemma is equal to another even if it has different orth order, or different pronunciation order. This is why it's set to `False` here.

Reply:
I think it depends on the use case. Or if you say, only `same_order=False` makes sense anyway, then why is this an option at all for `_equals`.

What actually is your use case?

Reply:
My use case is the following: in `MergeLexiconJob` I added functionality (present in this PR) that will remove "duplicated" lemmas. According to my definition, duplicated lemmas are those lemmas that have the same orthographies and pronunciations. Note that my definition implies no ordering, so I would be fine with `same_order=False`.

However, I wanted to give the freedom to let the user be able to decide whether the order in their orths and prons matters or not. Maybe some users might want to store the orths and prons of a lexicon in lexicographical order, and thus the order would matter. Maybe some other users already have a specific ordering, and therefore the addition of a lemma with a different order helps them notice that something's wrong with their pipeline.

Summarizing: `same_order=False` is enough for my needs/use case, but I added this flag since I was unsure of the _other_ use cases for this function, since each user might have a particular use case.

If you think it won't be used and only `==`/`!=` will be used as I have been using it (with `same_order=False`), let me know. I won't use this flag, and if keeping it or removing it will spark a debate, I'd rather just remove it now and let another user, who might find it more necessary, implement it.

Reply:
So, if you think there is no clear non-ambiguous definition of an equal-relation, thus having such flag makes sense, then I'm not sure if defining `__eq__` makes sense here. Or usually `__eq__` should always be most restrictive, i.e. use `same_order=True`.

Now `__eq__` is very arbitrary, supporting exactly your case, but as you say yourself, there might be other cases where the equality should only be with `same_order=True`.

Reply:
I agree, `__eq__` matches my use case but it doesn't cover the most restrictive use case.

What do others think? Is it too overkill to have `same_order` to test lemma equality? Should we just have the case of "two lemmas are equal whenever they contain the same data, regardless of their order"?

At every comment I'm more inclined to remove `same_order` and just check for what I said above, since I think no realistic use case would consider the ordering of data inside a lemma... At least I can't think of any specific use case for this...

Reply:
In general, we should not add code/logic for cases which are only hypothetical and not used currently, but only for things we really are using currently. Thus only the logic `same_order=False`. And then `__eq__` is also non-ambiguous.

But let's wait what other think about this.